modules/file-input/main/index.coffee

```diff
@@ -0,0 +1,165 @@
+        .add image
+        .add removeButton
+
+    getFileUID = (file) -> file.name + file.size + file.lastModified + file.type
+
+    handleFiles = (fileMap) =>
+      selectedFiles.clear()
+        .classed 'hx-btn', false
+        .off('click', 'hx.file-input')
+      length = fileMap.size
+
+      if length
+        if length is 1
+          dropdown.hide() if dropdown.isOpen()
+          selectedFiles.append filePreview(fileMap.values()[0])
+        else
+          selectedFiles
+            .classed 'hx-btn', true
+            .add hx.detached('span').text "#{length} Files Selected"
+            .add hx.detached('i').class 'hx-file-input-dropdown-icon hx-icon hx-icon-chevron-down'
+            .on 'click', 'hx.file-input', -> dropdown.show()
+      else
+        selectedFiles.append 'span'
+          .text resolvedOptions.noFilesText or ''
+
+      @_.fileMap = fileMap
+
+      @emit 'change',
+        cause: 'user'
+        data: @_.fileMap.values()
+
+    fileListToMap = (fileList) ->
+      # Deals with duplicates if the file has the same UID
+      map = new hx.Map()
+      for file in fileList
+        fileUID = getFileUID file
+        map.set(fileUID, file)
+      map
+
+    if resolvedOptions.multiple and not options.buttonText
+      resolvedOptions.buttonText = 'Choose Files'
+
+    selection = hx.select @selector
+      .classed 'hx-file-input', true
+
+    input = hx.detached 'input'
+      .class 'hx-file-input-hidden'
+      .attr 'type', 'file'
+      .attr 'accept', resolvedOptions.accept
+      .attr 'multiple', if resolvedOptions.multiple then 'multiple' else undefined
+
+    group = hx.inputGroup()
+
+    button = hx.detached 'button'
+      .class "hx-file-input-button hx-btn #{resolvedOptions.buttonClass}"
+      .on 'click', -> input.node().click()
+      .add hx.detached('i').class('hx-file-input-icon hx-icon hx-icon-upload')
+      .add hx.detached('span').text resolvedOptions.buttonText
+
+    selectedFiles = hx.detached 'div'
+      .class 'hx-file-input-selected hx-section'
+      .text resolvedOptions.noFilesText or ''
+
+    dropdownDiv = hx.detached 'div'
+    setupDropdown = (element) ->
+      sel = hx.select element
+      self._.fileMap.values().map (file) ->
+        sel.append filePreview(file)
+    dropdown = new hx.Dropdown(dropdownDiv.node(), setupDropdown, {
+      ddClass: 'hx-file-input-dropdown'
+    })
+
+    input.on 'change', (e) -> handleFiles fileListToMap e.target.files
+
+    @_ =
+      options: resolvedOptions
+      input: input
+      button: button
+      fileMap: new hx.Map()
+
+    if resolvedOptions.dragEnabled
+      preventDefault = (e) ->
+        e.preventDefault()
+        e.stopPropagation()
+
+      drop = (e) =>
+        preventDefault(e)
+        unless @_.disabled
+          files = e.dataTransfer.files
+          validates = true
+          for file in files
+            break if not validates
```

Comment:
This validation loop currently does nothing - I'm investigating how to do this whilst retaining the current 'accepts' property of the file input